debugger/vscode-dap-extension/src/starlingMonkeyDebugger.ts

```diff
@@ -0,0 +1,384 @@
+        new StoppedEvent("step", StarlingMonkeyDebugSession.threadID)
+      );
+    });
+    this._runtime.on("stopOnBreakpoint", () => {
+      this.sendEvent(
+        new StoppedEvent("breakpoint", StarlingMonkeyDebugSession.threadID)
+      );
+    });
+    this._runtime.on("stopOnDataBreakpoint", () => {
+      this.sendEvent(
+        new StoppedEvent("data breakpoint", StarlingMonkeyDebugSession.threadID)
+      );
+    });
+    this._runtime.on("stopOnInstructionBreakpoint", () => {
+      this.sendEvent(
+        new StoppedEvent(
+          "instruction breakpoint",
+          StarlingMonkeyDebugSession.threadID
+        )
+      );
+    });
+    this._runtime.on("stopOnException", (exception) => {
+      if (exception) {
+        this.sendEvent(
+          new StoppedEvent(
+            `exception(${exception})`,
+            StarlingMonkeyDebugSession.threadID
+          )
+        );
+      } else {
+        this.sendEvent(
+          new StoppedEvent("exception", StarlingMonkeyDebugSession.threadID)
+        );
+      }
+    });
+    this._runtime.on("output", (type, text, filePath, line, column) => {
+      let category: string;
+      switch (type) {
+        case "prio":
+          category = "important";
+          break;
+        case "out":
+          category = "stdout";
+          break;
+        case "err":
+          category = "stderr";
+          break;
+        default:
+          category = "console";
+          break;
+      }
+      const e: DebugProtocol.OutputEvent = new OutputEvent(
+        `${text}\n`,
+        category
+      );
+
+      if (text === "start" || text === "startCollapsed" || text === "end") {
+        e.body.group = text;
+        e.body.output = `group-${text}\n`;
+      }
+
+      e.body.source = this.createSource(filePath);
+      e.body.line = this.convertDebuggerLineToClient(line);
+      e.body.column = this.convertDebuggerColumnToClient(column);
+      this.sendEvent(e);
+    });
+    this._runtime.on("end", () => {
+      this.sendEvent(new TerminatedEvent());
+    });
+  }
+
+  protected initializeRequest(
+    response: DebugProtocol.InitializeResponse,
+    _args: DebugProtocol.InitializeRequestArguments
+  ): void {
+    response.body = {
+      ...response.body,
+      supportsSetVariable: true,
+      supportsConfigurationDoneRequest: true,
+      supportsBreakpointLocationsRequest: true,
+    };
+
+    this.sendResponse(response);
+  }
+
+  protected configurationDoneRequest(
+    response: DebugProtocol.ConfigurationDoneResponse,
+    args: DebugProtocol.ConfigurationDoneArguments
+  ): void {
+    super.configurationDoneRequest(response, args);
+    this._runtime.run();
+  }
+
+  protected disconnectRequest(
+    _response: DebugProtocol.DisconnectResponse,
+    args: DebugProtocol.DisconnectArguments,
+    _request?: DebugProtocol.Request
+  ): void {
+    console.log(
+      `disconnectRequest suspend: ${args.suspendDebuggee}, terminate: ${args.terminateDebuggee}`
+    );
```

Comment:
Should this send the response as well?

Reply:
There's nothing to do: the JS DAP library equips us with a "default" response (line 151), and there's no need for us to modify anything about it (https://microsoft.github.io/debug-adapter-protocol/specification#Requests_Disconnect, final paragraph).  And the DAP library takes care of the actual sending.